locksmith/src/controllers/hookController.ts

```diff
@@ -0,0 +1,203 @@
+      callback: hub.callback,
+      mode: hub.mode,
+      secret: hub.secret,
+      network: params.network,
+      lock: params.lock,
+    })
+
+    return createdHook
+  }
+
+  if (hook.mode !== hub.mode) {
+    hook.mode = hub.mode
+  }
+
+  if (hub.lease_seconds) {
+    hook.expiration = getExpiration(hub.lease_seconds)
+  }
+
+  if (hub.secret) {
+    hook.secret = hub.secret
+  }
+
+  const check = await isValidHubIntent(hub)
+
+  if (!check) {
+    throw new Error(
+      `Subscriber failed to confirm intent for callback: ${hub.callback} on topic: ${hub.topic}.`
+    )
+  }
+
+  await hook.save()
+  return hook
+}
+
+// Check if the subscriber request is valid or not. This will post a challenge to subscriber to confirm the intent of the request.
+export async function isValidHubIntent(hub: z.infer<typeof Hub>) {
+  try {
+    const challenge = crypto.randomBytes(20).toString('hex')
+    const result = await fetch(hub.callback, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify({
+        hub: {
+          ...hub,
+          challenge,
+        },
+      }),
+    })
+
+    const json = await result.json()
+
+    if (!result.ok) {
+      throw new Error("Subscriber didn't confirm intent.")
+    }
+
+    if (json.challenge !== challenge) {
+      throw new Error('Challenge did not match')
+    }
+
+    if (json.topic !== hub.topic) {
+      throw new Error('Topic did not match')
+    }
+
+    if (json.callback !== hub.callback) {
+      throw new Error('Callback did not match')
+    }
+
+    return hub
+  } catch (error) {
+    logger.error(error.message)
+    return false
+  }
+}
+
+export function getSupportedNetwork(network: string) {
+  if (!networks[network]) {
+    logger.error(`Unsupported network: ${network}`)
+    return
+  }
+  return networks[network]
+}
+
+export async function subscriptionHandler(req: Request, res: Response) {
+  try {
+    // We check the hub schema here to make sure the subscriber is sending us the correct data.
+    const hub = await Hub.parseAsync(req.body.hub)
+    // We check the network here to make sure the subscriber is sending to the right network endpoint.
+    const network = getSupportedNetwork(req.params.network)
+    if (!network) {
+      return res.status(400).send('Unsupported network')
+    }
+    // We send the accepted request to the subscriber and then validate the intent of the subscriber as well as persist the subscription.
+    res.status(202).send('Accepted')
+    await subscribe(hub, req.params)
+    return
+  } catch (error) {
+    logger.error(error)
+    // We respond with the error if we cannot subscribe or there is an error in the subscriber request.
+    return res.status(500).send({
+      hub: {
+        mode: req.body.hub.mode,
+        topic: req.body.hub.topic,
+        reason: error.message,
+      },
+    })
+  }
+}
+
+export function publisherHandler(req: Request, res: Response) {
```

Comment:
One thing though, should we verify if the lock address they are trying to get is valid. It would be a request to the subgraph and throw if the lock param is present.

Reply:
We could eventually but I think for now it is ok to go without it.